tests/suite/reconfig_test.go

```diff
@@ -0,0 +1,614 @@
+		if err != nil {
+			return 0, err
+		}
+
+		return int(parsedTS2.Sub(parsedTS1).Seconds()), nil
+	}
+
+	calculateTimeToReadyAverage := func(ngfLogs string) (string, error) {
+		var reconcilingLine, nginxReloadLine string
+		const maxCount = 5
+
+		var times [maxCount]int
+		var count int
+
+		// parse the logs until it reaches a reconciling log line for a gateway resource, then it compares that
+		// timestamp to the next NGINX configuration update. When it reaches the NGINX configuration update line,
+		// it will reset the reconciling log line and set it to the next reconciling log line.
+		for _, line := range strings.Split(ngfLogs, "\n") {
+			// can't just do this line, need to do gateway specific resources
+			if reconcilingLine == "" &&
+				strings.Contains(line, "Reconciling the resource\",\"controller\"") &&
+				strings.Contains(line, "\"controllerGroup\":\"gateway.networking.k8s.io\"") {
+				reconcilingLine = line
+			}
+
+			if strings.Contains(line, "NGINX configuration was successfully updated") && reconcilingLine != "" {
+				nginxReloadLine = line
+
+				timeDifference, err := calculateTimeDifferenceBetweenLogLines(reconcilingLine, nginxReloadLine)
+				if err != nil {
+					return "", err
+				}
+				reconcilingLine = ""
+
+				times[count] = timeDifference
+				count++
+				if count == maxCount-1 {
+					break
+				}
+			}
+		}
+
+		var sum float64
+		for _, time := range times {
+			sum += float64(time)
+		}
+
+		avgTime := sum / float64(count+1)
+
+		if avgTime < 1 {
+			return "< 1", nil
+		}
+
+		return strconv.FormatFloat(avgTime, 'f', -1, 64), nil
+	}
+
+	calculateTimeToReadyTotal := func(ngfLogs, startingLogSubstring string) (string, error) {
+		var firstLine, lastLine string
+		for _, line := range strings.Split(ngfLogs, "\n") {
+			if firstLine == "" && strings.Contains(line, startingLogSubstring) {
+				firstLine = line
+			}
+
+			if strings.Contains(line, "NGINX configuration was successfully updated") {
+				lastLine = line
+			}
+		}
+
+		timeToReadyTotal, err := calculateTimeDifferenceBetweenLogLines(firstLine, lastLine)
+		if err != nil {
+			return "", err
+		}
+
+		stringTimeToReadyTotal := strconv.Itoa(timeToReadyTotal)
+		if stringTimeToReadyTotal == "0" {
+			stringTimeToReadyTotal = "< 1"
+		}
+
+		return stringTimeToReadyTotal, nil
+	}
+
+	runTestWithMetrics := func(
+		testName string,
+		resourceCount int,
+		test func(resourceCount int) error,
+		startWithNGFSetup bool,
+		timeToReadyStartingLogSubstring string,
+	) {
+		var (
+			metricExistTimeout = 2 * time.Minute
+			metricExistPolling = 1 * time.Second
+			ngfPodName         string
+			startTime          time.Time
+		)
+
+		getStartTime := func() time.Time { return startTime }
+		modifyStartTime := func() { startTime = startTime.Add(500 * time.Millisecond) }
+
+		output, err := framework.InstallGatewayAPI(getDefaultSetupCfg().gwAPIVersion)
+		Expect(err).ToNot(HaveOccurred(), string(output))
+
+		if startWithNGFSetup {
```

Comment:
Anyone able to see a way to refactor this code of the if statement followed by the running of the test, then the other if statement? Can't think of anything that would provide benefit much better than the current.

Reply:
What about moving the setup steps into the test cases and only keeping the common test steps in runTestWithMetrics:


```
When("Resources exists before NGF is created", func(){
	It("Reconfigures with 30 resources", func(){
		Expect(createResourcesGWLast(30)).To(Succeed())
		Expect(checkResourceCreation(30)).To(Succeed())

		setup(getDefaultSetupCfg())
		podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
		Expect(err).ToNot(HaveOccurred())
		Expect(podNames).To(HaveLen(1))
		ngfPodName = podNames[0]

		runTestWithMetrics("Test 1", ngfPodName, logSubstring)
	})
})
		
When("NGF is installed before resources are created", func(){
	It("Reconfigures with 30 resources", func(){
		setup(getDefaultSetupCfg())
		podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
		Expect(err).ToNot(HaveOccurred())
		Expect(podNames).To(HaveLen(1))
		ngfPodName = podNames[0]
		
		Expect(createResourcesRoutesLast(30)).To(Succeed())
		Expect(checkResourceCreation(30)).To(Succeed())

		runTestWithMetrics("Test 2", ngfPodName, logSubstring)
		})
	})
```

You could also extract the setup NGF steps into a function that returned the ngfPodName.

Reply:
Did some refactoring, unsure of the name of the function, but otherwise I think that what I have is what I like. Please let me know of your thoughts!

Reply:
LGTM!